shared-bindings/supervisor/__init__.c

```diff
@@ -311,6 +312,63 @@ STATIC mp_obj_t supervisor_reset_terminal(mp_obj_t x_pixels, mp_obj_t y_pixels)
 }
 MP_DEFINE_CONST_FUN_OBJ_2(supervisor_reset_terminal_obj, supervisor_reset_terminal);
 
+//| def set_usb_descriptor_overrides(manufacturer: Optional[str] = None, product: Optional[str] = None, vid: Optional[int] = None, pid: Optional[int] = None) -> None:
+//|     """Override constants in the USB Device Descriptor.
+//|
+//|     Any omitted arguments will be left at their current or default values.
+//|     This method must be called in boot.py to have any effect."""
+//|     ...
+//|
+STATIC mp_obj_t supervisor_set_usb_descriptor_overrides(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_manufacturer, MP_ARG_OBJ, {.u_rom_obj = mp_const_none} },
+        { MP_QSTR_product, MP_ARG_OBJ, {.u_rom_obj = mp_const_none} },
+        { MP_QSTR_vid, MP_ARG_INT, {.u_int = -1} },
+        { MP_QSTR_pid, MP_ARG_INT, {.u_int = -1} },
+    };
+    struct {
+        mp_arg_val_t manufacturer;
+        mp_arg_val_t product;
+        mp_arg_val_t vid;
+        mp_arg_val_t pid;
+    } args;
+    mp_arg_parse_all(n_args, pos_args, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, (mp_arg_val_t *)&args);
+
+    if (!mp_obj_is_str_or_bytes(args.manufacturer.u_obj) && args.manufacturer.u_obj != mp_const_none) {
+        mp_raise_TypeError(translate("argument has wrong type"));
+    }
+    if (!mp_obj_is_str_or_bytes(args.product.u_obj) && args.product.u_obj != mp_const_none) {
+        mp_raise_TypeError(translate("argument has wrong type"));
+    }
+
+    const char *manufacturer = NULL;
+    const char *product = NULL;
+
+    size_t len;
+    if (args.manufacturer.u_obj != mp_const_none) {
+        manufacturer = mp_obj_str_get_data(args.manufacturer.u_obj, &len);
+        if (len > 126) {
+          mp_raise_ValueError(translate("string too long"));
+        }
+    }
+    if (args.product.u_obj != mp_const_none) {
+        product = mp_obj_str_get_data(args.product.u_obj, &len);
+        if (len > 126) {
+          mp_raise_ValueError(translate("string too long"));
+        }
+    }
+    if (args.vid.u_int < -1 || args.vid.u_int >= (1 << 16)) {
+        mp_raise_ValueError(translate("vid or pid out of range"));
+    }
+    if (args.pid.u_int < -1 || args.pid.u_int >= (1 << 16)) {
+        mp_raise_ValueError(translate("vid or pid out of range"));
```

Comment:
Use `mp_get_buffer_raise()` here to check for `Union[str, bytes, ReadableBuffer]`, instead of checking for `str` or `bytes`. For examples, see `shared-bindings/wifi/Radio.c`. That will take care of the type checking. Validate the length using `mp_arg_validate_length_range()`, in `py/argcheck.c` Search for examples of its use. Then you don't need to introduce new error message strings. We try not to include new strings unless we have to.

As you probably know, the descriptor strings are actually sent as UTF-16 values, but we don't really support that currrently. One can specify multiple languages. That is a big complication and we don't provide that either; it would be a lot of extra code.

For the VID and PID validation, use `mp_arg_validate_int_range()`/

Reply:
Thanks for the detailed pointers. I thought as much about error strings, but didn't find the right tools right away. `mp_arg_validate_*` makes this much cleaner.

I now realize that I approached the UTF-16 problem the wrong way, the buffers do need to be 255 bytes wide (as a maximum) and ASCII text needs to be padded into UTF-16 theoretically. In my testing I didn't do that and it still worked fine, but obviously we shouldn't rely on that.

Ideally, the function would take `str` or `bytes` and:
- encode strs of length < 127 into UTF-16
- use `bytes` of length < 255 as-is

But the code for building the descriptor string t able (`usb_build_interface_string_table`) is shared with other descriptors right now and does ASCII -> UTF-16 conversion. So I think limiting the input to ASCII for now is an acceptable middle ground.

Reply:
I agree about limiting it to ASCII; that's not a problem.